napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,463 @@
+            _add_sample,
+            plugin=mf.name,
+            sample=sample.key,
+        )
+
+        if multiprovider:
+            title = sample.display_name
+        else:
+            title = menu_item_template.format(
+                mf.display_name, sample.display_name
+            )
+        # To display '&' instead of creating a shortcut
+        title = title.replace('&', '&&')
+
+        action: Action = Action(
+            id=f'{mf.name}:{sample.key}',
+            title=title,
+            menus=[{'id': submenu_id, 'group': MenuGroup.NAVIGATION}],
+            callback=_add_sample_partial,
+        )
+        sample_actions.append(action)
+    return submenu, sample_actions
+
+
+def _get_widget_viewer_param(
+    widget_callable: WidgetCreator, widget_name: str
+) -> str:
+    """Get widget parameter name for `viewer` (if any) and check type."""
+    if inspect.isclass(widget_callable) and issubclass(
+        widget_callable,
+        (QWidget, Widget),
+    ):
+        widget_param = ''
+        try:
+            sig = inspect.signature(widget_callable.__init__)
+        except ValueError:  # pragma: no cover
+            # Inspection can fail when adding to bundled viewer as it thinks widget is
+            # a builtin
+            pass
+        else:
+            for param in sig.parameters.values():
+                if param.name == 'napari_viewer' or param.annotation in (
+                    'napari.viewer.Viewer',
+                    Viewer,
+                ):
+                    widget_param = param.name
+                    break
+
+    # For magicgui type widget contributions, `Viewer` injection is done by
+    # `magicgui.register_type`.
+    elif isinstance(widget_callable, MagicFactory) or inspect.isfunction(
+        widget_callable
+    ):
+        widget_param = ''
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _toggle_or_get_widget(
+    plugin: str,
+    widget_name: str,
+    full_name: str,
+) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+    """Toggle if widget already built otherwise return widget.
+
+    Returned widget will be added to main window by a processor.
+    Note for magicgui type widget contributions, `Viewer` injection is done by
+    `magicgui.register_type` instead of a provider via annnotation.
+    """
+    viewer = _provide_viewer_or_raise(
+        msg='Note that widgets cannot be opened in headless mode.',
+    )
+
+    window = viewer.window
+    if window and (dock_widget := window._dock_widgets.get(full_name)):
+        dock_widget.setVisible(not dock_widget.isVisible())
+        return None
```

Comment:
This should be tested. It is an important use case. (If not in this PR proper issue should be created first).

Reply:
Some discussion previously here https://github.com/napari/napari/pull/4991#discussion_r1368056355
Opened an issue: #6722

Reply:
Maybe it is meant to be tested a different way/using different function? Let me know if you have solution.

Reply:
Responded with solution in issue